### PR TITLE
[v2.5] Re-add support for getambassador.io/v1

### LINKED ISSUE #4240: K8s control plane logs conversion webhook errors
**Describe the bug**
After upgrading from Ambassador API gateway 1.13 to Emissary-ingress 2.2.2, the control plane logs have started showing the following errors continuously:
```
E0519 17:40:52.693825      12 cacher.go:419] cacher (*unstructured.Unstructured): unexpected ListAndWatch error: failed to list getambassador.io/v1, Kind=Mapping: conversion webhook for getambassador.io/v2, Kind=Mapping failed: no kind "Mapping" is registered for version "getambassador.io/v1" in scheme "/go/cmd/apiext/main.go:50"; reinitializing...
E0519 17:40:52.538414      12 cacher.go:419] cacher (*unstructured.Unstructured): unexpected ListAndWatch error: failed to list getambassador.io/v1, Kind=AuthService: conversion webhook for getambassador.io/v2, Kind=AuthService failed: no kind "AuthService" is registered for version "getambassador.io/v1" in scheme "/go/cmd/apiext/main.go:50"; reinitializing...
E0519 17:40:52.332153      12 cacher.go:419] cacher (*unstructured.Unstructured): unexpected ListAndWatch error: failed to list getambassador.io/v1, Kind=RateLimitService: conversion webhook for getambassador.io/v2, Kind=RateLimitService failed: no kind "RateLimitService" is registered for version "getambassador.io/v1" in scheme "/go/cmd/apiext/main.go:50"; reinitializing...
E0519 17:40:52.180641      12 cacher.go:419] cacher (*unstructured.Unstructured): unexpected ListAndWatch error: failed to list getambassador.io/v1, Kind=TCPMapping: conversion webhook for getambassador.io/v2, Kind=TCPMapping failed: no kind "TCPMapping" is registered for version "getambassador.io/v1" in scheme "/go/cmd/apiext/main.go:50"; reinitializing...
E0519 17:40:52.147282      12 cacher.go:419] cacher (*unstructured.Unstructured): unexpected ListAndWatch error: failed to list getambassador.io/v1, Kind=TracingService: conversion webhook for getambassador.io/v2, Kind=TracingService failed: no kind "TracingService" is registered for version "getambassador.io/v1" in scheme "/go/cmd/apiext/main.go:50"; reinitializing...
E0519 17:40:51.746911      12 cacher.go:419] cacher (*unstructured.Unstructured): unexpected ListAndWatch error: failed to list getambassador.io/v1, Kind=KubernetesEndpointResolver: conversion webhook for getambassador.io/v2, Kind=KubernetesEndpointResolver failed: no kind "KubernetesEndpointResolver" is registered for version "getambassador.io/v1" in scheme "/go/cmd/apiext/main.go:50"; reinitializing...
```

Besides the error logs, everything seems to work correctly.

**To Reproduce**
I don't know how to reproduce this issue

**Expected behavior**
I expect the control plane logs to not contain errors for getambassador.io/v1 objects.

**Versions (please complete the following information):**
 - Emissary-ingress: 2.2.2
 - Kubernetes AWS EKS v1.19.16-eks-25803e
 - Kubectl 1.23.5

**Additional context**
We only have 1 tracingservice in our clusters. 

First, it was a tracingservice version in v1 as an annotation
```
apiVersion: v1
kind: Service
metadata:
  name: zipkin
  namespace: monitoring
  annotations:
    getambassador.io/config: |
      ---
      apiVersion: ambassador/v1
      kind: TracingService
      name: tracing
      service: zipkin.monitoring:9411
      driver: zipkin
...
```  
We deleted the annotation and applied to the following v2 custom resource 1 year ago:
```
apiVersion: getambassador.io/v2
kind: TracingService
metadata:
  name: tracing
  namespace: monitoring
spec:
  ambassador_id:
    - internal
    - public
  driver: zipkin
  service: zipkin.monitoring:9411
  tag_headers:
    - :authority
    - :path
```

The number of error logs around 10700 log messages per 15 minutes, so its spams a lot.

Kubernetes CLI reports that there are stored versions of the CRD in v1 and v2 (see bottom lines):
```
$ kubectl get crd tracingservices.getambassador.io -o=yaml

apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  annotations:
    controller-gen.kubebuilder.io/version: v0.6.2
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"apiextensions.k8s.io/v1","kind":"CustomResourceDefinition","metadata":{"annotations":{"controller-gen.kubebuilder.io/version":"v0.6.2"},"labels":{"app.kubernetes.io/instance":"emissary-apiext","app.kubernetes.io/managed-by":"kubectl_apply_-f_emissary-apiext.yaml","app.kubernetes.io/name":"emissary-apiext","app.kubernetes.io/part-of":"emissary-apiext","owner":"team_sre"},"name":"tracingservices.getambassador.io"},"spec":{"conversion":{"strategy":"Webhook","webhook":{"clientConfig":{"service":{"name":"emissary-apiext","namespace":"ambassador"}},"conversionReviewVersions":["v1beta1"]}},"group":"getambassador.io","names":{"categories":["ambassador-crds"],"kind":"TracingService","listKind":"TracingServiceList","plural":"tracingservices","singular":"tracingservice"},"preserveUnknownFields":false,"scope":"Namespaced","versions":[{"name":"v2","schema":{"openAPIV3Schema":{"description":"TracingService is the Schema for the tracingservices API","properties":{"apiVersion":{"description":"APIVersion defines the versioned schema of this representation of an object. Servers should convert recognized schemas to the latest internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources","type":"string"},"kind":{"description":"Kind is a string value representing the REST resource this object represents. Servers may infer this from the endpoint the client submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds","type":"string"},"metadata":{"type":"object"},"spec":{"description":"TracingServiceSpec defines the desired state of TracingService","properties":{"config":{"properties":{"access_token_file":{"type":"string"},"collector_cluster":{"type":"string"},"collector_endpoint":{"type":"string"},"collector_endpoint_version":{"enum":["HTTP_JSON_V1","HTTP_JSON","HTTP_PROTO"],"type":"string"},"collector_hostname":{"type":"string"},"service_name":{"type":"string"},"shared_span_context":{"type":"boolean"},"trace_id_128bit":{"type":"boolean"}},"type":"object"},"driver":{"enum":["lightstep","zipkin","datadog"],"type":"string"},"sampling":{"properties":{"client":{"type":"integer"},"overall":{"type":"integer"},"random":{"type":"integer"}},"type":"object"},"service":{"type":"string"},"tag_headers":{"items":{"type":"string"},"type":"array"},"v3StatsName":{"type":"string"}},"required":["driver","service"],"type":"object","x-kubernetes-preserve-unknown-fields":true}},"type":"object"}},"served":true,"storage":true},{"name":"v3alpha1","schema":{"openAPIV3Schema":{"description":"TracingService is the Schema for the tracingservices API","properties":{"apiVersion":{"description":"APIVersion defines the versioned schema of this representation of an object. Servers should convert recognized schemas to the latest internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources","type":"string"},"kind":{"description":"Kind is a string value representing the REST resource this object represents. Servers may infer this from the endpoint the client submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds","type":"string"},"metadata":{"type":"object"},"spec":{"description":"TracingServiceSpec defines the desired state of TracingService","properties":{"ambassador_id":{"description":"AmbassadorID declares which Ambassador instances should pay attention to this resource. If no value is provided, the default is: \n    ambassador_id:    - \"default\" \n TODO(lukeshu): In v3alpha2, consider renaming all of the `ambassador_id` (singular) fields to `ambassador_ids` (plural).","items":{"type":"string"},"type":"array"},"config":{"properties":{"access_token_file":{"type":"string"},"collector_cluster":{"type":"string"},"collector_endpoint":{"type":"string"},"collector_endpoint_version":{"enum":["HTTP_JSON_V1","HTTP_JSON","HTTP_PROTO"],"type":"string"},"collector_hostname":{"type":"string"},"service_name":{"type":"string"},"shared_span_context":{"type":"boolean"},"trace_id_128bit":{"type":"boolean"}},"type":"object"},"driver":{"enum":["lightstep","zipkin","datadog"],"type":"string"},"sampling":{"properties":{"client":{"type":"integer"},"overall":{"type":"integer"},"random":{"type":"integer"}},"type":"object"},"service":{"type":"string"},"stats_name":{"type":"string"},"tag_headers":{"items":{"type":"string"},"type":"array"}},"required":["driver","service"],"type":"object"}},"type":"object"}},"served":true,"storage":false},{"name":"v1","schema":{"openAPIV3Schema":{"description":"TracingService is the Schema for the tracingservices API","type":"object","x-kubernetes-preserve-unknown-fields":true}},"served":false,"storage":false}]}}
  creationTimestamp: "2019-09-12T09:15:35Z"
  generation: 12
  labels:
    app.kubernetes.io/instance: emissary-apiext
    app.kubernetes.io/managed-by: kubectl_apply_-f_emissary-apiext.yaml
    app.kubernetes.io/name: emissary-apiext
    app.kubernetes.io/part-of: emissary-apiext
    owner: team_sre
  name: tracingservices.getambassador.io
  resourceVersion: "2436618420"
  selfLink: /apis/apiextensions.k8s.io/v1/customresourcedefinitions/tracingservices.getambassador.io
  uid: e124e532-d53d-11e9-b8d3-0ab5b0c37404
spec:
  conversion:
    strategy: Webhook
    webhook:
      clientConfig:
        caBundle: <redacted>
        service:
          name: emissary-apiext
          namespace: ambassador
          path: /webhooks/crd-convert
          port: 443
      conversionReviewVersions:
      - v1beta1
  group: getambassador.io
  names:
    categories:
    - ambassador-crds
    kind: TracingService
    listKind: TracingServiceList
    plural: tracingservices
    singular: tracingservice
  scope: Namespaced
  versions:
  - name: v2
    schema:
      openAPIV3Schema:
        description: TracingService is the Schema for the tracingservices API
        properties:
          apiVersion:
            description: 'APIVersion defines the versioned schema of this representation
              of an object. Servers should convert recognized schemas to the latest
              internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources'
            type: string
          kind:
            description: 'Kind is a string value representing the REST resource this
              object represents. Servers may infer this from the endpoint the client
              submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds'
            type: string
          metadata:
            type: object
          spec:
            description: TracingServiceSpec defines the desired state of TracingService
            properties:
              config:
                properties:
                  access_token_file:
                    type: string
                  collector_cluster:
                    type: string
                  collector_endpoint:
                    type: string
                  collector_endpoint_version:
                    enum:
                    - HTTP_JSON_V1
                    - HTTP_JSON
                    - HTTP_PROTO
                    type: string
                  collector_hostname:
                    type: string
                  service_name:
                    type: string
                  shared_span_context:
                    type: boolean
                  trace_id_128bit:
                    type: boolean
                type: object
              driver:
                enum:
                - lightstep
                - zipkin
                - datadog
                type: string
              sampling:
                properties:
                  client:
                    type: integer
                  overall:
                    type: integer
                  random:
                    type: integer
                type: object
              service:
                type: string
              tag_headers:
                items:
                  type: string
                type: array
              v3StatsName:
                type: string
            required:
            - driver
            - service
            type: object
            x-kubernetes-preserve-unknown-fields: true
        type: object
    served: true
    storage: true
  - name: v3alpha1
    schema:
      openAPIV3Schema:
        description: TracingService is the Schema for the tracingservices API
        properties:
          apiVersion:
            description: 'APIVersion defines the versioned schema of this representation
              of an object. Servers should convert recognized schemas to the latest
              internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources'
            type: string
          kind:
            description: 'Kind is a string value representing the REST resource this
              object represents. Servers may infer this from the endpoint the client
              submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds'
            type: string
          metadata:
            type: object
          spec:
            description: TracingServiceSpec defines the desired state of TracingService
            properties:
              ambassador_id:
                description: "AmbassadorID declares which Ambassador instances should
                  pay attention to this resource. If no value is provided, the default
                  is: \n    ambassador_id:    - \"default\" \n TODO(lukeshu): In v3alpha2,
                  consider renaming all of the `ambassador_id` (singular) fields to
                  `ambassador_ids` (plural)."
                items:
                  type: string
                type: array
              config:
                properties:
                  access_token_file:
                    type: string
                  collector_cluster:
                    type: string
                  collector_endpoint:
                    type: string
                  collector_endpoint_version:
                    enum:
                    - HTTP_JSON_V1
                    - HTTP_JSON
                    - HTTP_PROTO
                    type: string
                  collector_hostname:
                    type: string
                  service_name:
                    type: string
                  shared_span_context:
                    type: boolean
                  trace_id_128bit:
                    type: boolean
                type: object
              driver:
                enum:
                - lightstep
                - zipkin
                - datadog
                type: string
              sampling:
                properties:
                  client:
                    type: integer
                  overall:
                    type: integer
                  random:
                    type: integer
                type: object
              service:
                type: string
              stats_name:
                type: string
              tag_headers:
                items:
                  type: string
                type: array
            required:
            - driver
            - service
            type: object
        type: object
    served: true
    storage: false
  - name: v1
    schema:
      openAPIV3Schema:
        description: TracingService is the Schema for the tracingservices API
        type: object
        x-kubernetes-preserve-unknown-fields: true
    served: false
    storage: false
status:
  acceptedNames:
    categories:
    - ambassador-crds
    kind: TracingService
    listKind: TracingServiceList
    plural: tracingservices
    singular: tracingservice
  conditions:
  - lastTransitionTime: "2019-09-12T09:15:35Z"
    message: no conflicts found
    reason: NoConflicts
    status: "True"
    type: NamesAccepted
  - lastTransitionTime: null
    message: the initial names have been accepted
    reason: InitialNamesAccepted
    status: "True"
    type: Established
  storedVersions:
  - v1
  - v2
```

Kubectl only reports a single tracingservice and it is in v2:
```
$ kubectl get tracingservices.getambassador.io --all-namespaces -o=yaml

apiVersion: v1
items:
- apiVersion: getambassador.io/v2
  kind: TracingService
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"getambassador.io/v2","kind":"TracingService","metadata":{"annotations":{},"name":"tracing","namespace":"monitoring"},"spec":{"ambassador_id":["public"],"driver":"zipkin","service":"zipkin.monitoring:9411","tag_headers":[":authority",":path"]}}
    creationTimestamp: "2021-03-03T22:19:56Z"
    generation: 2
    name: tracing
    namespace: monitoring
    resourceVersion: "791986493"
    selfLink: /apis/getambassador.io/v2/namespaces/monitoring/tracingservices/tracing
    uid: c69f7e0f-d621-4d52-b644-7bca965fba5f
  spec:
    ambassador_id:
    - public
    driver: zipkin
    service: zipkin.monitoring:9411
    tag_headers:
    - :authority
    - :path
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

### PR DESCRIPTION
**Pull request by LukeShu**
## Description

There are 3 reasons why this PR is worth-while (ordered most-compelling first):
 1. To resolve #4240; we'll discuss that below.
 2. To port-over the daisy-chain conversion logic from Edge Stack, which will be valuable when we move past apiVersion `v3alpha`.  I discuss this in more length at https://github.com/emissary-ingress/emissary/wiki/design-doc:-Emissary's-CRD-conversion-logic
 3. To remove friction for users migrating from Emissary 1.x.

### Why this fixes #4240 

Background: Users coming from Emissary 1.x likely have resources stored as `getambassador.io/v1`.  That is, the CRDs' `status.storedVersions` will include `v1`.  In order to be able to even _apply_ the newer `crds.yaml`, it must include all of the old `status.storedVersions`.  So, the 2.1-2.4 `crds.yaml` includes an "unserved" v1 that is unusable by users, but satisfies Kubernetes' requirement that all versions in `status.storedVersions` be mentioned in the definition. (Thanks to @kflynn for reminding me of this background--I'd totally forgotten it.)

Related: https://www.notion.so/datawire/smooth-upgrades-how-v1-is-handled-aae77e4dcb554a6bafcdfc071875b48f

However, it turns out that even though it's unserved, Kube wants to ask apiext about it--even if it's a fresh 2.x/3.x install. That's what #4240.  We try to could come up with ways to get Kube to stop asking about v1.  But it's fairly easy to just actually add support for v1 so that we don't error when Kube asks about it; plus this has the 2 other benefits that it reduces friction for users and that we'll benefit from the machinery in the future, so even if it were harder than the alternative, it would likely still be worth it.

## Related Issues
- https://github.com/emissary-ingress/emissary/issues/4240

## Testing
I've verified that I don't see the `no kind "XXX" is registered for version "getambassador.io/v1"` errors from `docker logs` when running in k3d that @kflynn said he sees.  But he should verify as well that this resolves things for his set up.

## Checklist

 - [x] I made sure to update `CHANGELOG.md`.
   
   Remember, the CHANGELOG needs to mention:
    + Any new features
    + Any changes to our included version of Envoy
    + Any non-backward-compatible changes
    + Any deprecations
 
 - [ ] This is unlikely to impact how Ambassador performs at scale. - I'd be shocked if any of the places I touch are in the hot-path, but there probably are more allocations now, so IDK
 
   Remember, things that might have an impact at scale include:
    + Any significant changes in memory use that might require adjusting the memory limits
    + Any significant changes in CPU use that might require adjusting the CPU limits
    + Anything that might change how many replicas users should use
    + Changes that impact data-plane latency/scalability
 
 - [ ] My change is adequately tested. - pending feedback from @kflynn 
 
   Remember when considering testing:
    + Your change needs to be specifically covered by tests.
       + Tests need to cover all the states where your change is relevant: for example, if you add a behavior that can be enabled or disabled, you'll need tests that cover the enabled case and tests that cover the disabled case. It's not sufficient just to test with the behavior enabled.
    + You also need to make sure that the _entire area being changed_ has adequate test coverage.
       + If existing tests don't actually cover the entire area being changed, add tests.
       + This applies even for aspects of the area that you're not changing – check the test coverage, and improve it if needed!
    + We should lean on the bulk of code being covered by unit tests, but...
    + ... an end-to-end test should cover the integration points
 
 - [x] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently. - no new tricks

### CODE REVIEW

**Review by LanceEa — COMMENTED** [review]
Thanks for tracking this down and addressing it so quickly. I will give it a deeper review on Monday, but at first glance it looks good. I left a few questions. 

I think it would be good to test out @kflynn reproducer to verify it isn't occurring still and we will want to make sure this gets pulled into Edge Stack right away verify it doesn't have any issues (_I don't think there will be_)

File: build-aux/generate.mk
Comment (LukeShu) [reply build-aux/generate.mk:324]: Correct.

File: pkg/api/getambassador.io/v2/handwritten.conversion.go
Comment (LukeShu) [reply pkg/api/getambassador.io/v2/handwritten.conversion.go:214]: Yeah, because the `controller-runtime` APIs don't have a "context" argument for us to provide.

**Review by LanceEa — COMMENTED** [review]
Overall looks good, a couple of comments and questions. 

1. @kflynn - it would be good to have you verify it
2. Do we need to pull this into Edge Stack and verify all is good before merging? My initial reaction is no because these are isolated to the `apiext` so we shouldn't have any issues.

**Review by kflynn — APPROVED** [review]
Looks good from here, and I can verify that it shuts up the logs for me, too. I think that @LanceEa is correct about the mostly-static scaffold file, but I think that's a good topic for further discussion and a future PR, rather than a necessary thing right now.

Anecdotally, I saw timeouts from apiext when testing this, which worries me a bit. Time to look into that, maybe.

File: build-aux/generate.mk
Comment (LukeShu) [reply build-aux/generate.mk:324]: But you would need to manually copy it between apiversions.

File: pkg/api/getambassador.io/v1/crd_devportal.go
Comment (LukeShu) [reply pkg/api/getambassador.io/v1/crd_devportal.go:40]: To put it somewhere public: from the private #emissary-maintainers channel: 

@kflynn:
> One of @LanceEa’s comments on #4055 has me wondering about `DevPortal` – do I remember correctly that it’s in the Emissary CRDs because way TF back when we didn’t have separate Edge Stack CRDs? Presumably we could move it now, right?

@LanceEA:
> Yeah, I haven’t played with this feature much but I had thought they were an EdgeStack only feature so what your saying makes sense to me if that is true.
>
> Does any of the generated routes in diagd look at these? If so then that is probably part of the isEdgeStack things that we need to address before we could move these.

@kflynn:
> AFAIK there is *literally* nothing in Emissary that understands `DevPortal` at all.
>
> This is definitely a place where I’d love to get confirmation from @AliceProxy and @LukeShu though.

@AliceProxy:
> That's also what I thought

@LukeShu:
> We *always* had a way for them to be separate.
>
> So I'm sure there has to be a reason it went in Emissary.

@kflynn:
> Searching for `DevPortal` in the Emissary directory, all I see are tests and a comment in `pkg/agent/api_docs.go` about “be careful because there’s a similar scraping implementation in DevPortal”.
>
> `DevPortals` do appear in the snapshot; is that so Cloud can see them? are the Edge Stack types not sent back to Cloud?

@AliceProxy:
> Probably for the cloud devportal stuff
>
> As far as I'm aware, anything in the snapshot should be sent to the cloud.

@kflynn:
> Well that’s too bad. :joy:

@LukeShu:
> The PR to blame: https://github.com/datawire/apro/pull/2007
>
> I think the answer is "Alvaro didn't know how to add AES-only CRDs, and Flynn didn't catch it in code review"
>
> No! I think the answer is "So that it could take advantage of the main watcher/snapshot, instead of having to implement another watcher in amb-sidecar"

@AliceProxy:
> That was my reason for moving Filters over to Interesting Types in emissary
>
> (And so you could process the secret refs in the reconciler steps)

@kflynn:
> I think #2 is the reason, yeah. Damn. I’d forgotten about that.

File: pkg/kates/internal/borrowed_webhook.go
Comment (LukeShu) [reply pkg/kates/internal/borrowed_webhook.go:46]: IDK, this is code copied ("borrowed") from controller-runtime.

File: build-aux/generate.mk
Comment (LukeShu) [reply build-aux/generate.mk:324]: I could move it to a `build-aux/conversion-spoke.go.gen` that looks like

```bash
#!/usr/bin/env bash
set -euE -o pipefail

[[ $# == 1 ]]
dir=$1

{
	cat <<EOT
package ${dir##*/}

import (
	k8sRuntime "k8s.io/apimachinery/pkg/runtime"
	"sigs.k8s.io/controller-runtime/pkg/conversion"
)

func convertFrom(src conversion.Hub, dst conversion.Convertible) error {
	scheme := conversionScheme()
	var cur k8sRuntime.Object = src
	for i := len(conversionIntermediates) - 1; i >= 0; i-- {
		gv := conversionIntermediates[i]
		var err error
		cur, err = scheme.ConvertToVersion(cur, gv)
		if err != nil {
			return err
		}
	}
	return scheme.Convert(cur, dst, nil)
}

func convertTo(src conversion.Convertible, dst conversion.Hub) error {
	scheme := conversionScheme()
	var cur k8sRuntime.Object = src
	for _, gv := range conversionIntermediates {
		var err error
		cur, err = scheme.ConvertToVersion(cur, gv)
		if err != nil {
			return err
		}
	}
	return scheme.Convert(cur, dst, nil)
}

EOT
	gawk '
	  BEGIN {
	    object=0;
	  }
	  /\/\/ \+kubebuilder:object:root=true/ {
	    object=1;
	  }
	  /^type \S+ struct/ && object {
	    if (!match($2, /List$/)) {
	      print "func(dst *" $2 ") ConvertFrom(src conversion.Hub) error { return convertFrom(src, dst) }";
	      print "func(src *" $2 ") ConvertTo(dst conversion.Hub) error { return convertTo(src, dst) }";
	    }
	    object=0;
	  }' \
	  "$dir"/*.go
} | gofmt

```

but I'm not actually sure that's much better

Comment (LukeShu) [reply build-aux/generate.mk:324]: I suppose there could be a `build-aux/conversion-spoke.head.go` or something with just that first snippet. But that first snippet is just an implementation convenience of the `ConvertFrom` and `ConvertTo` methods, spreading that across multiple files seems like it's making things harder, not easier.

Comment (LukeShu) [reply build-aux/generate.mk:324]: Maybe just moving them to stand-alone `.awk` files might be better, though.

**Comment by LukeShu** [comment]
2 changes:
- I've added a commit moving the AWK scripts to separate files.
- I've merged https://github.com/emissary-ingress/emissary/pull/4683 to fix the build

**Comment by LukeShu** [comment]
Rebased on to https://github.com/emissary-ingress/emissary/pull/4685 to resolve linter issues

**Review by LanceEa — DISMISSED** [review]
lgtm...Looks like this is landing in 2.5 so we will want to get it landed in master too.

**Comment by LukeShu** [comment]
Fixed the changelog:
```patch
diff --git a/CHANGELOG.md b/CHANGELOG.md
index 884539fa1..302918068 100644
--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -82,8 +82,8 @@ it will be removed; but as it won't be user-visible this isn't considered a brea
 
 ## RELEASE NOTES
 
-## [2.5.0] November 03, 2022
-[2.5.0]: https://github.com/emissary-ingress/emissary/compare/v2.4.0...v2.5.0
+## [2.5.1] TBD
+[2.5.1]: https://github.com/emissary-ingress/emissary/compare/v2.5.0...v2.5.1
 
 ### Emissary-ingress and Ambassador Edge Stack
 
@@ -94,6 +94,11 @@ it will be removed; but as it won't be user-visible this isn't considered a brea
   Kubernetes Nodes (regardless of whether the installation was migrated from 1.y or was a fresh 2.y
   install); fully supporting `v1` again should resolve these errors.
 
+## [2.5.0] November 03, 2022
+[2.5.0]: https://github.com/emissary-ingress/emissary/compare/v2.4.0...v2.5.0
+
+### Emissary-ingress and Ambassador Edge Stack
+
 - Bugfix: If a `Host` or `TLSContext` contained a hostname with a `:` then when using the 
   diagnostics endpoints `ambassador/v0/diagd` then an error would be thrown due to the parsing logic
   not  being able to handle the extra colon. This has been fixed and Emissary-ingress will not throw
diff --git a/docs/releaseNotes.yml b/docs/releaseNotes.yml
index 1c935bd74..a38f8fa9a 100644
--- a/docs/releaseNotes.yml
+++ b/docs/releaseNotes.yml
@@ -32,8 +32,8 @@
 
 changelog: https://github.com/emissary-ingress/emissary/blob/$branch$/CHANGELOG.md
 items:
-  - version: 2.5.0
-    date: '2022-11-03'
+  - version: 2.5.1
+    date: TBD
     notes:
       - title: Re-add support for getambassador.io/v1
         type: feature
@@ -46,6 +46,9 @@ items:
           installation was migrated from 1.y or was a fresh 2.y install); fully supporting
           <code>v1</code> again should resolve these errors.
 
+  - version: 2.5.0
+    date: '2022-11-03'
+    notes:
       - title: Diagnostics stats properly handles parsing envoy metrics with colons
         type: bugfix
         body: >-
```

**Review by LanceEa — COMMENTED** [review]
looks like there still is a merge conflict

**Comment by LukeShu** [comment]
Rebased to resolve merge conflict in the changelog.

**Review by LanceEa — APPROVED** [review]

**Review by AliceProxy — APPROVED** [review]
This looks ok to me

**Comment by joshbranham** [comment]
What is the timetable for a 2.5.x release with this fix? Would love to resolve the constant spewing of this log if possible.
```
cacher.go:424] cacher (*unstructured.Unstructured): unexpected ListAndWatch error: failed to list getambassador.io/v1, Kind=Module: conversion webhook for getambassador.io/v2, Kind=Module failed: no kind "Module" is registered for version "getambassador.io/v1" in scheme "/go/cmd/apiext/main.go:50"; reinitializing...
```

**Pull request merged**
